### ISSUE TITLE
Cannot compute gradients through variable reparameterizations

### ISSUE THREAD

**Issue by AtheMathmo** [pytorch/pytorch#3123]
Hi!

This is likely to be user error but I can't figure out why this fails and have reduced to a minimal example.

I want to reparameterize some parameters to restrict their domain. In particular I have some variances I want to optimize which must remain positive. So I create some new variables to parameterize the variances and optimize over these instead. In particular, I use,

```python
import torch
from torch.autograd import Variable

# rho is my dummy variable
rho_init = torch.ones(100)
rho = Variable(rho_init, requires_grad=True)
# Here the variances are reparameterized to always be positive
variances = torch.log(1.0 + torch.exp(rho))

# Do optimization
optimizer = torch.optim.Adam([rho], lr=0.01)

for i in range(10):
    optimizer.zero_grad()
    x = torch.sum(variances)
    x.backward()
    optimizer.step()
```

However this loop will fail when it hits the second `backward()` call. My guess is that rho is updated and so the variances must be recomputed but the computational graph is not discarded. It is not obvious to me whether this is intended behaviour. If this is expected, how should I go about making use of such reparameterizations?

The only issue I could find which may be related is #1605 but this seems to be resolved.

**Comment by AtheMathmo** [comment]
Closed as I realised my mistake. I need to update the variances manually after each rho update from the optimizer.
